### PR TITLE
ci: introduce upgrade tests (WIP)

### LINKED ISSUE #126: Add firmware update scenario to CI
To catch issues like https://github.com/trezor/trezor-firmware/issues/122 which do not appear unless the storage has the format of a previous firmware release.

Once https://github.com/trezor/trezor-firmware/issues/125 is implemented we can cache emulator binaries of older firmware builds and perform tests across two versions.

### PR DESCRIPTION
**Pull request by tsusanka**
This introduces upgrade tests. The Emulator Wrapper and python script were written by @prusnak originally with some modifications of mine. The emulators are built on CI and stored on our [internal dev server](https://firmware.corp.sldev.cz/upgrade_tests/).

All versions that are found on the website are tested. I will also add a CI job, which will deploy new emulators to this website if a new tag occurs (meaning anytime we tag a new version, it will be deployed there automatically and therefore upgrade-tested). This closes #126.

----

We need to:

- [x] Review the Python part.
- [x] Do we want to call this using pytest? That's why other tests are failing currently, because pytest picks up the files as well (they have test_* functions). The upgrade tests themselves are [passing](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/277110968) though.
- [x] Solve the MINIMUM VERSION [override](https://github.com/trezor/trezor-firmware/compare/tsusanka/ci-upgrade-tests2?expand=1#diff-a51b8f647ebcb623f5181189925e48c1L5). I wasn't sure how to do it properly. 
- [x] @prusnak We need to build the emulator frozen with PYOPT=0. [This](https://github.com/trezor/trezor-firmware/compare/tsusanka/ci-upgrade-tests2?expand=1#diff-84be10005b1d7d80ca912486c46f07d8R48) is a bit hackish - I guess we could add a make target for this, right? And I guess I need to change SConscript.unix for that, because it is overwritten otherwise.
- [ ] Test if some "upgrade bug" is introduced, this actaully catches it.

@matejcik do you think I could leave this one for you to finish? Might be more effective than doing the review ping-ping.

### CODE REVIEW

**Comment by prusnak** [comment]
I added the missing functionality in b8726e74c81cf19aff9cfa9e0886d268e97471ee

Just use the following in the CI:

```
build core unix frozen debug:
  stage: build
  variables:
    PYOPT=0
  script:
    - cd core
    - pipenv run make build_unix_frozen
```

**Comment by tsusanka** [comment]
> I added the missing functionality in b8726e7

Added!

**Comment by matejcik** [comment]
I modified the tests to run with pytest, and modified the other pytest runners accordingly. I also fixed the minversion override.

The tests are currently passing (for me); the only problem is that the reset test is failing on `core` because it needs swiping.
Thinking of which... I don't know how it can be passing on `legacy`, because there is no code to input the reset words?

File: tests/upgrade_tests/test_firmware_upgrades.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+#         print("[%s] %s => %s" % (gen, tag_from, tag_to))
+#         print("- test_upgrade_load")
+#         test_upgrade_load(gen, tag_from, tag_to)
+#         if gen != "core":
+#             print("- test_upgrade_reset")
+#             test_upgrade_reset(gen, tag_from, tag_to)
+#             print("- test_upgrade_reset_skip_backup")
+#             test_upgrade_reset_skip_backup(gen, tag_from, tag_to)
+#             print("- test_upgrade_reset_no_backup")
+#             test_upgrade_reset_no_backup(gen, tag_from, tag_to)
```
Comment (tsusanka) [inline tests/upgrade_tests/test_firmware_upgrades.py]: Let's remove this?

**Comment by tsusanka** [comment]
> the only problem is that the reset test is failing on core because it needs swiping

I suggest to merge this now and create an issue for such extension.

File: tests/upgrade_tests/test_firmware_upgrades.py (hunk 1 above)
Comment (matejcik) [reply tests/upgrade_tests/test_firmware_upgrades.py]: done

**Comment by tsusanka** [comment]
Squashed and rebased. Merging

**Pull request merged**